### PR TITLE
Issue #1780 - Project source/destination removal should consider wildcards

### LINKED ISSUE #1780: Project destination removal does not consider wildcards
We had a project which had some project destinations, e.g.

```yaml
- server: https://server1.com
  namespace: my-service-qa
- server: https://server2.com
  namespace: my-service-prod
```

Using the API, we tried to simplify the project by collapsing this using wildcards:
```yaml
- server: *
  namespace: my-service-*
```

However, the API server validation failed with error `Unable to edit project: following destinations are used by one or more application and cannot be removed`.

The bug is that the API server check does not consider wildcards when checking if a project destination removal is already covered by some wildcard destination.

The workaround was to manually edit the project (e.g. using kubectl) to remove the undesired destinations.

### PR DESCRIPTION
**Pull request by alexmt**
Closes #1780

### CODE REVIEW

File: server/project/project_test.go [hunk 1]
```diff
@@ -193,6 +193,50 @@ func TestProjectServer(t *testing.T) {
...
+		projectServer := NewServer("default", fake.NewSimpleClientset(), apps.NewSimpleClientset(proj, &existingApp), enforcer, util.NewKeyLock(), nil)
+
+		updatedProj := proj.DeepCopy()
+		updatedProj.Spec.Destinations = []v1alpha1.ApplicationDestination{
+			{Namespace: "org1-*", Server: "https://server1"},
+		}
+
+		_, err := projectServer.Update(context.Background(), &project.ProjectUpdateRequest{Project: updatedProj})
+
+		assert.Nil(t, err)
```
Comment (alexec) [inline server/project/project_test.go]: minor `NoError`

File: server/project/project_test.go [hunk 2]
```diff
@@ -193,6 +193,50 @@ func TestProjectServer(t *testing.T) {
...
+		}
+
+		projectServer := NewServer("default", fake.NewSimpleClientset(), apps.NewSimpleClientset(proj, &existingApp), enforcer, util.NewKeyLock(), nil)
+
+		updatedProj := proj.DeepCopy()
+		updatedProj.Spec.SourceRepos = []string{"https://github.com/argoproj/*"}
+
+		_, err := projectServer.Update(context.Background(), &project.ProjectUpdateRequest{Project: updatedProj})
+
+		assert.Nil(t, err)
```
Comment (alexec) [inline server/project/project_test.go]: minor - `NoError`

Comment (alexec) [reply server/project/project_test.go]: I feel like you should be doing more asserts here?

File: server/project/project_test.go (hunk 1 above)
Comment (alexec) [reply server/project/project_test.go]: I feel like you should be doing more asserts here too.

File: server/project/project.go [hunk 3]
```diff
@@ -256,31 +216,41 @@ func (s *Server) Update(ctx context.Context, q *project.ProjectUpdateRequest) (*
...
-	if len(removedSrcUsed) > 0 {
-		return nil, status.Errorf(
-			codes.InvalidArgument, "following source repos are used by one or more application and cannot be removed: %s", strings.Join(removedSrcUsed, ";"))
+	for _, a := range dstValidatedApps {
+		if !q.Project.IsDestinationPermitted(a.Spec.Destination) {
+			invalidDstCount++
+		}
+	}
+
+	if invalidSrcCount > 0 || invalidDstCount > 0 {
```
Comment (alexec) [inline server/project/project.go]: minor, but you could declare parts, append, and then check to see if `len(parts) > 0` to throw error

**Review by alexec — APPROVED** [review]

File: server/project/project.go (hunk 3 above)
Comment (alexmt) [reply server/project/project.go]: good suggestion. done

File: server/project/project_test.go (hunk 2 above)
Comment (alexmt) [reply server/project/project_test.go]: applied suggestion. also added assert that project actually updated

File: server/project/project_test.go (hunk 1 above)
Comment (alexmt) [reply server/project/project_test.go]: done

**Comment by alexec** [comment]
Pro-actively merging.

**Pull request merged**